### PR TITLE
network_put and network_get modules

### LINKED ISSUE #39238: [FILE_FETCH] network_put module to copy large config file to target device with scp/sftp
##### ISSUE TYPE
 - Feature Idea

##### ANSIBLE VERSION

```
2.6
```

##### SUMMARY
Sometimes the provided configuration are too large and devices will take a long time to accept them, error out, or need timeout value modification (or all of the above).It could be helpful in certain situations to allow the user to use scp or sftp to copy the configuration file over.

This module does not take care of pre-confiuguration required on device to use scp. It also does not issue any replace config etc command after copying config file. It is a generic file transfer module. pre and post command of file copy should be handled in playbook using [os]_command specific module

##### STEPS TO REPRODUCE

```yaml

```

##### ACTUAL RESULTS

```

```

### PR DESCRIPTION
**Pull request by gdpak**
##### SUMMARY
Fixes #39238 

##### ISSUE TYPE
  - New Module Pull Request

##### ANSIBLE VERSION
2.6

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure due to import error:

```
ImportError while importing test module '/root/ansible/test/units/plugins/cliconf/test_slxos.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
test/units/plugins/cliconf/test_slxos.py:28: in <module>
    from ansible.plugins.cliconf import slxos
lib/ansible/plugins/cliconf/__init__.py:30: in <module>
    import q
E   ImportError: No module named q
```

File: lib/ansible/modules/network/files/network_get.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  src:
+    description:
+      - path of the source file.
+    required: true
+  protocol:
+    description:
+      - protocol used to transfer file
+    default: scp
+    choices: ['scp', 'sftp']
+    required: no
```
Comment (gundalow) [inline lib/ansible/modules/network/files/network_get.py]: `required: no` is the default and can be removed.
This comment applies to both modules

File: lib/ansible/modules/network/files/network_get.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+version_added: "2.6"
+author: "Deepak Agrawal (@dagrawal)"
+short_description: copy files from network device to ansible controller
+description:
+  - This module provides functionlity to copy file from network device to
+    ansible controller.
+options:
+  src:
+    description:
+      - path of the source file.
```
Comment (gundalow) [inline lib/ansible/modules/network/files/network_get.py]: `Path of the source file.`
Absolute or relative?
If relative, relative to what?

File: lib/ansible/modules/network/files/network_get.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  protocol:
+    description:
+      - protocol used to transfer file
+    default: scp
+    choices: ['scp', 'sftp']
+    required: no
+  dest:
+    description:
+      - destination file name with absolute path
+    default: playbook or role root directory if playbook is part of a role
```
Comment (gundalow) [inline lib/ansible/modules/network/files/network_get.py]: `default:` need to match argspec
`playbook or role root directory if playbook is part of a role` should be added to `description:`

File: lib/ansible/modules/network/files/network_get.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    required: no
+
+requirements:
+    - "scp"
+
+notes:
+   - Some devices need specific configurations to be enabled before scp can work
+     These configuration should be pre-configued before using this module
+     e.g ios - ip scp server enable
```
Comment (gundalow) [inline lib/ansible/modules/network/files/network_get.py]: `C(ip scp server enable)`

File: lib/ansible/modules/network/files/network_get.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+   - User privileage to do scp on network device should be pre-configured
+     e.g. ios - need user privileage 15 by default for allowing scp
+   - Default destination of source file
+"""
+
+EXAMPLES = """
+- name: copy file from ios device to ansible controller
```
Comment (gundalow) [inline lib/ansible/modules/network/files/network_get.py]: Don't know that's it's ios, so
`- name: Copy file from the network device to Ansible controller`

File: lib/ansible/modules/network/files/network_get.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+DOCUMENTATION = """
+---
+module: network_get
+version_added: "2.6"
+author: "Deepak Agrawal (@dagrawal)"
+short_description: copy files from network device to ansible controller
+description:
+  - This module provides functionlity to copy file from network device to
+    ansible controller.
+options:
```
Comment (gundalow) [inline lib/ansible/modules/network/files/network_get.py]: For consistency, please ensure all `descriptions:` start with a capital letter, and end with a full stop.

File: lib/ansible/modules/network/files/network_get.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'network'}
+
+
+DOCUMENTATION = """
+---
+module: network_get
+version_added: "2.6"
+author: "Deepak Agrawal (@dagrawal)"
+short_description: copy files from network device to ansible controller
```
Comment (gundalow) [inline lib/ansible/modules/network/files/network_get.py]: `Copy files from a network device to Ansible controller`

File: lib/ansible/modules/network/files/network_put.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'network'}
+
+
+DOCUMENTATION = """
+---
+module: network_put
+version_added: "2.6"
+author: "Deepak Agrawal (@dagrawal)"
+short_description: copy files from ansibe controller to  network devices
```
Comment (gundalow) [inline lib/ansible/modules/network/files/network_put.py]: `Copy files from Ansible controller to a network device`
(single device)

File: lib/ansible/modules/network/files/network_put.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+description:
+  - This module provides functionlity to copy file from controller to
```
Comment (gundalow) [inline lib/ansible/modules/network/files/network_put.py]: from Ansible controller

File: lib/ansible/plugins/action/network_get.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
+# (c) 2015, Ansible Inc,
```
Comment (gundalow) [inline lib/ansible/plugins/action/network_get.py]: 2018?

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/action/network_get.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import copy
+import os
+import time
+import re
+import q
```
Comment (ganeshrn) [inline lib/ansible/plugins/action/network_get.py]: `import q` need to be removed

File: lib/ansible/plugins/action/network_get.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+            return {'failed': True, 'msg': 'missing required argument: %s' % exc}
+
+        # Get destination file if specified
+        dest = self._task.args.get('dest')
+
+        if dest is None:
+            dest = self._get_default_dest(src)
+        else:
+            dest = self._handle_dest_path(dest)
+            q(dest)
```
Comment (ganeshrn) [inline lib/ansible/plugins/action/network_get.py]: This lined should be removed

File: lib/ansible/plugins/action/network_get.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+        # Get proto
+        proto = self._task.args.get('protocol')
+        if proto is None:
+            proto = 'scp'
+
+        sock_timeout = play_context.timeout
```
Comment (ganeshrn) [inline lib/ansible/plugins/action/network_get.py:71]: Based on PR https://github.com/ansible/ansible/pull/39367 the timeout value need to be fetched from get_option
```
self.get_option('persistent_command_timeout')

```

Comment (gdpak) [reply lib/ansible/plugins/action/network_get.py:71]: Sure I will change it once I this change is available in devel.

**Comment by gdpak** [comment]
@gundalow @ganeshrn review changes are incorporated. Please have a look

**Review by ganeshrn — APPROVED** [review]

**Comment by mattclay** [comment]
Should the modules be named `net_put` and `net_get` for consistency?

All of the existing platform agnostic modules have the `net_` prefix in their names:

- interface/net_lldp_interface.py
- interface/net_interface.py
- interface/net_linkagg.py
- protocol/net_lldp.py
- system/net_banner.py
- system/net_logging.py
- system/net_user.py
- system/net_ping.py
- system/net_system.py
- layer2/net_vlan.py
- layer2/net_l2_interface.py
- layer3/net_l3_interface.py
- layer3/net_vrf.py
- routing/net_static_route.py

**Comment by gdpak** [comment]
@mattclay new network platform agnostic modules are named like network_* - network_command, network_config, network_get ... https://github.com/ansible/community/wiki/Network:Core-roadmap-2.6

**Comment by gundalow** [comment]
Lets discuss the naming convention in the team meeting today

**Comment by gdpak** [comment]
ok I can take care of name change if any in another pr

**Pull request merged**